include/rnp/rnp2.h

```diff
@@ -326,6 +330,251 @@ rnp_result_t rnp_generate_key_json(rnp_ffi_t ffi, const char *json, char **resul
+ *
+ * @param op pointer to opaque key generation context.
+ * @param hash string, representing the hash algorithm. See the rnp_op_generate_set_hash()
+ *             function description for the list of possible values.
+ * @return RNP_SUCCESS or error code if failed.
+ */
+rnp_result_t rnp_op_generate_add_pref_hash(rnp_op_generate_t op, const char *hash);
+
+/** Clear the preferred hash algorithms list, so default ones will be used.
+ *
+ * @param op pointer to opaque key generation context.
+ * @return RNP_SUCCESS or error code if failed.
+ */
+rnp_result_t rnp_op_generate_clear_pref_hashes(rnp_op_generate_t op);
+
+/** Add preferred compression algorithm to user preferences.
+ *  Note: the first added algorithm has the highest priority, then the second and so on.
+ *        Applicable only for the primary key generation.
+ *
+ * @param op pointer to opaque key generation context.
+ * @param compression string, representing the compression algorithm. Possible values are:
+ *                    "zip", "zlib", "bzip2"
+ * @return RNP_SUCCESS or error code if failed.
+ */
+rnp_result_t rnp_op_generate_add_pref_compression(rnp_op_generate_t op,
+                                                  const char *      compression);
+
+/** Clear the preferred compression algorithms list, so default ones will be used.
+ *
+ * @param op pointer to opaque key generation context.
+ * @return RNP_SUCCESS or error code if failed.
+ */
+rnp_result_t rnp_op_generate_clear_pref_compression(rnp_op_generate_t op);
+
+/** Add preferred encryption algorithm to user preferences.
+ *  Note: the first added algorithm has the highest priority, then the second and so on.
+ *        Applicable only for the primary key generation.
+ *
+ * @param op pointer to opaque key generation context.
+ * @param cipher string, representing the encryption algorithm.
+ *               See the rnp_op_generate_set_protection_cipher() function description for
+ *               the list of possible values.
+ * @return RNP_SUCCESS or error code if failed.
+ */
+rnp_result_t rnp_op_generate_add_pref_cipher(rnp_op_generate_t op, const char *cipher);
+
+/** Clear the preferred encryption algorithms list, so default ones will be used.
+ *
+ * @param op pointer to opaque key generation context.
+ * @return RNP_SUCCESS or error code if failed.
+ */
+rnp_result_t rnp_op_generate_clear_pref_ciphers(rnp_op_generate_t op);
+
+/** Set the preferred key server. Applicable only for the primary key.
+ *
+ * @param op pointer to opaque key generation context.
+ * @param keyserver NULL-terminated string with key server's URL, or NULL to delete it from
+ *                  user preferences.
+ * @return RNP_SUCCESS or error code if failed.
+ */
+rnp_result_t rnp_op_generate_set_pref_keyserver(rnp_op_generate_t op, const char *keyserver);
+
+/** Execute the prepared key or subkey generation operation.
+ *  Note: if you set protection algorithm, then you need to specify ffi password provider to
```

Comment:
I don't think we should use the password provider for this, we should allow setting it directly with something like `rnp_op_generate_set_protection_password`. That would be more consistent with `rnp_key_protect`, `rnp_key_unprotect`, `rnp_op_encrypt_add_password`, etc.

I really intended for the pass provider to be used in cases where a password *may* be required, but it is not certain. In this case, if the user is specifying a protection cipher/etc, a password is *definitely* required.

Reply:
Agree, at least we could add this method but fallback to password provider if required. Will fix this.